drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c

```diff
@@ -0,0 +1,1669 @@
+		v4l2_dbg(1, debug, &dev->v4l2_dev,
+			 "%s: Waiting for buffers to be returned - %d outstanding\n",
+			 __func__, atomic_read(&port->buffers_with_vpu));
+		ret = wait_for_completion_timeout(&dev->frame_cmplt, HZ);
+		if (ret <= 0) {
+			v4l2_err(&dev->v4l2_dev,
+				 "%s: Timeout waiting for buffers to be returned - %d outstanding\n",
+				 __func__,
+				 atomic_read(&port->buffers_with_vpu));
+			break;
+		}
+	}
+
+	/*
+	 * Release the VCSM handle here as otherwise REQBUFS(0) aborts because
+	 * someone is using the dmabuf before giving the driver a chance to do
+	 * anything about it.
+	 */
+	for (i = 0; i < q->num_buffers; i++) {
+		struct vb2_v4l2_buffer *vb2 = to_vb2_v4l2_buffer(q->bufs[i]);
+		struct bcm2835_isp_buffer *buf =
+			container_of(vb2, struct bcm2835_isp_buffer, vb);
+		bcm2835_isp_mmal_buf_cleanup(&buf->mmal);
+	}
+
+	atomic_dec(&dev->num_streaming);
+	/* If all ports disabled, then disable the component */
+	if (atomic_read(&dev->num_streaming) == 0) {
+		ret = vchiq_mmal_component_disable(dev->mmal_instance,
+						   dev->component);
+		if (ret) {
+			v4l2_err(&dev->v4l2_dev,
+				 "%s: Failed disabling component, ret %d\n",
+				 __func__, ret);
+		}
+	}
+
+	/*
+	 * Simply wait for any vb2 buffers to finish. We could take steps to
+	 * make them complete more quickly if we care, or even return them
+	 * ourselves.
+	 */
+	vb2_wait_for_all_buffers(&node->queue);
+}
+
+static const struct vb2_ops bcm2835_isp_node_queue_ops = {
+	.queue_setup		= bcm2835_isp_node_queue_setup,
+	.buf_init		= bcm2835_isp_buf_init,
+	.buf_prepare		= bcm2835_isp_buf_prepare,
+	.buf_queue		= bcm2835_isp_node_buffer_queue,
+	.buf_cleanup		= bcm2835_isp_buffer_cleanup,
+	.start_streaming	= bcm2835_isp_node_start_streaming,
+	.stop_streaming		= bcm2835_isp_node_stop_streaming,
+};
+
+static struct bcm2835_isp_fmt *get_default_format(struct bcm2835_isp_node *node)
+{
+	return &node->supported_fmts.list[0];
+}
+
+static inline unsigned int get_bytesperline(int width,
+					    struct bcm2835_isp_fmt *fmt)
+{
+	return ALIGN((width * fmt->depth) >> 3, fmt->bytesperline_align);
+}
+
+static inline unsigned int get_sizeimage(int bpl, int width, int height,
+					 struct bcm2835_isp_fmt *fmt)
+{
+	return (bpl * height * fmt->size_multiplier_x2) >> 1;
+}
+
+static int bcm2835_isp_s_ctrl(struct v4l2_ctrl *ctrl)
+{
+	struct bcm2835_isp_dev *dev =
+	      container_of(ctrl->handler, struct bcm2835_isp_dev, ctrl_handler);
+	struct bcm2835_isp_node *node = &dev->node[0];
+	int ret = 0;
+
+	/*
+	 * The ISP firmware driver will ensure these settings are applied on
+	 * a frame boundary, so we are safe to write them as they come in.
+	 *
+	 * Note that the bcm2835_isp_* param structures are identical to the
+	 * mmal-parameters.h definitions.  This avoids the need for unnecessary
+	 * field-by-field copying between structures.
+	 */
+	switch (ctrl->id) {
+	case V4L2_CID_RED_BALANCE:
+		dev->r_gain = ctrl->val;
+		if (set_wb_gains(node))
```

Comment:
All of these setters don't set up ret based on the return code, so rely on the initialiser above.
Is there gain in replacing with
```
int ret;
switch(ctrl->id)
{
case V4L2_CID_RED_BALANCE:
   dev->r_gain = ctrl->val;
   ret = set_wb_gain(node);
   break;
...
default:
   ret = -EINVAL;
}
if (ret)
			v4l2_err(&dev->v4l2_dev, "%s: Failed setting ctrl  \"%s\" (%08x), err %d\n",
				 __func__, ctrl->name, ctrl->id, ret);
```

Reply:
Done.  Does make the code look a bit neater.

Reply:
Had to make a slight change here:

```
if (ret) {
   v4l2_err(&dev->v4l2_dev, "%s: Failed setting ctrl  \"%s\" (%08x), err %d\n",
				 __func__, ctrl->name, ctrl->id, ret);
   ret = -EIO
}
```

Otherwise v4l2-compliance complains that the mmal return error code is invalid.  -EIO is acceptable but does flag as a warning.

The reason for the mmal failure is the LS parameters are invalid (all zeros) and it correctly rejects it.  Given this is a compound control and cannot have any min/max/default, I feel v4l2-compliance should not expect arbitrary data filled into the control to actually pass...

Reply:
Yup, one of the cleanups to do at some point is to get rid of all the VCHIQ error numbers, and map MMAL errors to kernel ones. Adding ```ret = -EIO``` is fine. -EINVAL is probably not allowed as the ctrl handler is supposed to have sanity checked most of the parameters anyway.

I'm assuming you're only seeing a compliance failure rather than lockup. Yes, it seems to be an oversight in compliance if it tries testing compound controls that can't have been sanity checked.